File: Leet-Codes/moveArray.py
numb = [1, 2, 3, 4, 5, 6, 7]

# 1. Função auxiliar que apenas inverte uma fatia do array in-place
def reverse_array(arr, start, end):
    while start < end:
        temporary_save = arr[start]
        arr[start] = arr[end]
        arr[end] = temporary_save
        start += 1
        end -= 1


# 2. Função principal que gerencia o fluxo da rotação
def moviment_array(numb, k):
    k = k % len(numb)  # Trata k caso seja maior que o tamanho da lista

    # As três chamadas de inversão:
    reverse_array(numb, 0, len(numb) - 1)
    reverse_array(numb, 0, k - 1)
    reverse_array(numb, k, len(numb) - 1)

    return numb


# Executando:
moviment_array(numb, 5)

print(numb)  # Resultado: [3, 4, 5, 6, 7, 1, 2]

File: Leet-Codes/moveArray.js
/*Dado um array nums e um inteiro k, rotacione o array para a direita em k passos in-place.
Ex: nums = [1,2,3,4,5,6,7], k = 3 → [5,6,7,1,2,3,4].*/

var numb = [1, 2, 3, 4, 5, 6, 7];

function reverseArray(arr, start, end) {
    while (start < end) {
        var temporarySave = arr[start];
        arr[start] = arr[end];
        arr[end] = temporarySave;
        start++;
        end--;
    }
}


function movimentArray(numb, k) {
    k = k % numb.length;


    reverseArray(numb, 0, numb.length - 1);
    reverseArray(numb, 0, k - 1);
    reverseArray(numb, k, numb.length - 1);

    return numb;
}


movimentArray(numb, 5);

console.log(numb); 
